File: 0102/bfs.cpp
#include <cassert>
#include <iostream>
#include <limits>
#include <queue>
#include <vector>

#include "leetcode/tree.hpp"

using namespace std;
using namespace LeetCode;

class Solution {
 public:
  vector<vector<int>> levelOrder(TreeNode* root) {
    vector<vector<int>> ret;
    if (!root) return ret;
    queue<pair<TreeNode*, int>> q;
    q.push({root, 0});
    pair<TreeNode*, int> node_pair;
    TreeNode* node;
    int level;
    while (!q.empty()){
      node_pair = q.front();
      q.pop();
      node = node_pair.first;
      level = node_pair.second;
      if (ret.size() <= level) ret.resize(level+1);
      ret[level].push_back(node->val);
      if (node->left) q.push({node->left, level+1});
      if (node->right) q.push({node->right, level+1});
    }
    return ret;
  }
};

int main() {
  Solution sol;
  TreeNode* node;
  vector<vector<int>> got;
  vector<int> expected;

  node = CreateTreeNode("3 9 20 # # 15 7 ;");
  got = sol.levelOrder(node);
  expected = {3};
  assert (got[0] == expected);
  expected = {9,20};
  assert (got[1] == expected);
  expected = {15,7};
  assert (got[2] == expected);
  DeleteTreeNode(node);

  node = CreateTreeNode("1 2 3 4 5 6 7 ;");
  got = sol.levelOrder(node);
  expected = {1};
  assert (got[0] == expected);
  expected = {2,3};
  assert (got[1] == expected);
  expected = {4,5,6,7};
  assert (got[2] == expected);
  DeleteTreeNode(node);

  got = sol.levelOrder(nullptr);
  assert(got.size() == 0);
}

File: 0404/bfs.py
from collections import deque

from leetcode.tree import String2Tree, TreeNode


class Solution:
    def sumOfLeftLeaves(self, root: TreeNode) -> int:
        ret = 0
        if root is None:
            return ret
        q = deque()
        q.append((root, False))
        while len(q) > 0:
            node, is_left = q.popleft()
            if node.left is None and node.right is None and is_left:
                ret += node.val
                continue
            if node.left is not None:
                q.append((node.left, True))
            if node.right is not None:
                q.append((node.right, False))
        return ret


if __name__ == '__main__':
    sol = Solution()

    #   3
    #  / \
    # 9  20
    #   /  \
    #  15   7
    root = String2Tree("3 9 20 # # 15 7")
    expected = 24
    assert sol.sumOfLeftLeaves(root) == expected

    #     3
    #    / \
    #   9  20
    #  /  /  \
    # 10 15   7
    root = String2Tree("3 9 20 10 # 15 7")
    expected = 25
    assert sol.sumOfLeftLeaves(root) == expected

    root = String2Tree("3")
    expected = 0
    assert sol.sumOfLeftLeaves(root) == expected

    root = None
    expected = 0
    assert sol.sumOfLeftLeaves(root) == expected
